File: python/0506.py
# Link: https://leetcode.com/problems/relative-ranks/description/
# Runtime: 72 ms
# Memory: 17.94 MB


class Solution:
    def findRelativeRanks(self, score: List[int]) -> List[str]:
        sorted_scores = []
        for index, result in enumerate(score):
            sorted_scores.append((-result, index))

        heapq.heapify(sorted_scores)

        ranks = [0] * len(score)
        place = 1
        while sorted_scores:
            original_index = heapq.heappop(sorted_scores)[1]
            if place == 1:
                ranks[original_index] = "Gold Medal"
            elif place == 2:
                ranks[original_index] = "Silver Medal"
            elif place == 3:
                ranks[original_index] = "Bronze Medal"
            else:
                ranks[original_index] = str(place)

            place += 1

        return ranks

File: c++/0506.cpp
// Link: https://leetcode.com/problems/relative-ranks/description/
// Runtime: 0 ms
// Memory: 13.75 MB

class Solution {
public:
  std::vector<std::string> findRelativeRanks(const std::vector<int> &score) {
    std::priority_queue<std::pair<int, int>> sortedScores;
    for (int i = 0; i < static_cast<int>(score.size()); i++)
      sortedScores.push({score[i], i});

    std::vector<std::string> ranks(score.size());
    int place = 1;
    while (!sortedScores.empty()) {
      int originalIndex = sortedScores.top().second;
      if (place == 1)
        ranks[originalIndex] = "Gold Medal";
      else if (place == 2)
        ranks[originalIndex] = "Silver Medal";
      else if (place == 3)
        ranks[originalIndex] = "Bronze Medal";
      else
        ranks[originalIndex] = std::to_string(place);

      sortedScores.pop();
      place++;
    }

    return ranks;
  }
};
